src/groth16/prove.rs

```diff
@@ -212,59 +211,288 @@ fn compute_h(d_vec: &mut DeviceVec<ScalarField>, coset_gen: Option<ScalarField>,
+            mul_scalars(
+                &batched_d_vec[start_idx + nof_coef * 2..end_idx],
+                &d_keys[..],
+                &mut batched_d_vec_copy[start_idx + nof_coef * 2..end_idx],
+                &cfg,
+            ).unwrap();
+        }
+    } else if let Some(coset_gen) = coset_gen {
+        ntt_cfg.coset_gen = coset_gen;
+    } else {
+        panic!("[compute_h_batched]: Neither coset_gen nor keys provided");
+    }
+    
+    // Inverse NTT (inverse = false)
+    icicle_ntt(&mut batched_d_vec, false, &ntt_cfg);
+
+    // Process each batch element: L * R - O
+    let mut d_h_results = Vec::with_capacity(batch_size);
+    for batch_idx in 0..batch_size {
+        let start_idx = batch_idx * d_vecs[0].len();
+        
+        mul_scalars(
+            &batched_d_vec[start_idx..start_idx + nof_coef], 
+            &batched_d_vec[start_idx + nof_coef..start_idx + nof_coef * 2], 
+            &mut batched_d_vec_copy[start_idx..start_idx + nof_coef], 
+            &cfg
+        ).unwrap();
+        
+        let mut d_h = DeviceVec::device_malloc(nof_coef).unwrap();
+        sub_scalars(
+            &batched_d_vec[start_idx..start_idx + nof_coef], 
+            &batched_d_vec[start_idx + nof_coef * 2..start_idx + nof_coef * 3], 
+            &mut d_h, 
+            &cfg
+        ).unwrap();
+        d_h_results.push(d_h);
+    }
+    
+    stream.synchronize().unwrap();
+    d_h_results
+}
+
+fn commitments(scalars: &[F], zkey: &ZKey, n_public: usize, stream: &IcicleStream) -> (ProjectiveG1, ProjectiveG1, ProjectiveG2, ProjectiveG1) {
+    let host_scalars = HostSlice::from_slice(scalars);
+    let mut msm_config = MSMConfig::default();
+    msm_config.is_async = false;
+    msm_config.c = 14;
+    
+    let a = commit_g1(&host_scalars[..], zkey, 5, "a", &msm_config, stream);
+    let b1 = commit_g1(&host_scalars[..], zkey, 6, "b1", &msm_config, stream);
+    let b = commit_g2(&host_scalars[..], zkey, 7, "b", &msm_config, stream);
+    let c = commit_g1(&host_scalars[n_public+1..], zkey, 8, "c", &msm_config, stream);
+
+    (a, b1, b, c)
+}
+
+fn commitments_batched(
+    scalars_array: &[Vec<F>], 
+    zkey: &ZKey, 
+    n_public: usize, 
+    stream: &IcicleStream,
+    batch_size: usize
+) -> Vec<(ProjectiveG1, ProjectiveG1, ProjectiveG2, ProjectiveG1)> {
+    
+    println!("commitments_batched: batch_size={}, n_public={}", batch_size, n_public);
+    println!("commitments_batched: scalars_array lengths: {:?}", scalars_array.iter().map(|s| s.len()).collect::<Vec<_>>());
+    
+    // Configure MSM for batch processing
+    let mut msm_config = MSMConfig::default();
+    msm_config.is_async = false;
+    msm_config.c = 14;
+    msm_config.batch_size = batch_size as i32;
+    msm_config.are_points_shared_in_batch = true; // All proofs use the same zkey
+    
+    // Concatenate all scalars for batch MSM
+    let total_scalars: Vec<F> = scalars_array.iter().flat_map(|s| s.iter()).cloned().collect();
```

Comment:
There shouldn't be any need to clone here
```suggestion
    let total_scalars: Vec<F> = scalars_array.into_iter().flatten().collect();
```

Reply:
This change doesn't compile. Let's keep it until better times.